CONTRIBUTING.md

```diff
@@ -456,6 +456,188 @@ checks if contributed code does not have the appropriate license headers.
+
+`TRACE` logs will normally only make sense when read alongside the code, and
+typically they will be read as a whole sequence of messages rather than in
+isolation.
+
+Even though `TRACE` logs may be very verbose, you should still exercise some
+judgement when deciding when to use them. In many cases it will be easier to
+understand the behaviour of the system using tests or by analysing the code
+itself rather than by trawling through hundreds of trivial log messages.
+
+It may not be easy, or even possible, to obtain `TRACE` logs from a production
+system. Therefore they are not appropriate for information that you would
+expect to be useful in diagnosing problems in production.
+
+##### `DEBUG`
+
+This is the next least verbose level and is also disabled by default. The
+target audience of this level typically comprises users or developers who are
+trying to diagnose an unexpected problem in a production system, perhaps to
+help determine whether a fault lies within Elasticsearch or elsewhere.
+
+Users should expect to be able to enable `DEBUG` logging on their production
+systems for a whole subsystem for an extended period of time without
+overwhelming the system or filling up their disks with logs, so it is important
+to limit the volume of messages logged at this level. On the other hand, these
+messages must still provide enough detail to diagnose the sorts of problems
+that Elasticsearch might encounter. In some cases it works well to collect
+information over a period of time and then log a complete summary, rather than
+recording every step of a process in its own message.
+
+It's possible that the reader of `DEBUG` logs is also reading the code, but
+that is less likely than for `TRACE` logs. You should strive to avoid
+terminology that only makes sense when reading the code, and also aim for
+messages at this level to be self-contained rather than intending them to be
+read as a sequence.
+
+It's often useful to log exceptions and other deviations from the "happy path"
+at `DEBUG` level. Exceptions logged at `DEBUG` should generally include the
+complete stack trace.
+
+##### `INFO`
+
+This is the next least verbose level, and the first level that is enabled by
+default. It is appropriate for recording important events in the life of the
+cluster, such as an index being created or deleted, or a snapshot starting or
+completing. Users will mostly ignore log messages at `INFO` level, but may use
+these messages to construct a high-level timeline of events leading up to an
+incident.
+
+This level is enabled by default so its target audience is the general
+population of users and administrators. You should use user-facing terminology
+and ensure that messages at this level are self-contained. In general you
+shouldn't log unusual events, particularly exceptions with stack traces, at
+`INFO` level. If the event is relatively benign then use `DEBUG`, whereas if
+the user should be notified then use `WARN`.
+
```

Comment:
I wonder whether it's worth commenting here that some readers are overly enthusiastic at interpreting info (or even debug/trace) messages as errors. Thus, it is important not to make the message sound like an error condition if it is not intended to be one.

```suggestion

It is often the case that people who read these logs are trying to determine why
their node is not behaving the way they expect - that is, they're in trouble
shooting mode. This can lead them to read more into a log message than the
developer intended. If your log message sounds like an error, then some
administrators will interpret it as one, even if it is logged at `INFO` level.
Where possible, `INFO` messages should prefer softer language such
as _"Did not find ..."_ rather than "Failed to find ..."_.

```

Reply:
Thanks, that's a great point. I added (approximately) this in 6942b041d2a13741434689ca599f49e811ceb858